test/instrumentation/modules/@aws-sdk/fixtures/use-client-sns.js

```diff
@@ -0,0 +1,216 @@
+    endpoint: snsClient.config.endpoint,
+    topicName,
+    region,
+  });
+
+  let command;
+  let data;
+
+  // https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/sns/command/ListTopicsCommand/
+  // for testing purposes, shouldn't be instrumented
+  command = new ListTopicsCommand({});
+  data = await snsClient.send(command);
+  assert(
+    apm.currentSpan === null,
+    'SNS span (or its HTTP span) should not be currentSpan after awaiting the task',
+  );
+  log.info({ data }, 'listTopics');
+
+  let topicArn;
+  const preexistingTopic = data.Topics.find(
+    (t) => t.TopicArn.split(':').pop() === topicName,
+  );
+
+  if (!preexistingTopic) {
+    // https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/sns/command/CreateTopicCommand/
+    // for testing purposes, shouldn't be instrumented
+    command = new CreateTopicCommand({ Name: topicName });
+    data = await snsClient.send(command);
+    assert(
+      apm.currentSpan === null,
+      'SNS span (or its HTTP span) should not be currentSpan after awaiting the task',
+    );
+    log.info({ data }, 'createTopic');
+    topicArn = data.TopicArn;
+  } else {
+    topicArn = preexistingTopic.TopicArn;
+  }
+
+  // https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/sns/command/PublishCommand/
+  command = new PublishCommand({
+    Message: 'message to be sent',
+    PhoneNumber: '+34555555555',
+  });
+  data = await snsClient.send(command);
+  assert(
+    apm.currentSpan === null,
+    'SNS span (or its HTTP span) should not be currentSpan after awaiting the task',
+  );
+  log.info({ data }, 'publish with PhoneNumber');
+
+  command = new PublishCommand({
+    Message: 'message to be sent',
+    TopicArn: topicArn,
+  });
+  data = await snsClient.send(command);
+  assert(
+    apm.currentSpan === null,
+    'SNS span (or its HTTP span) should not be currentSpan after awaiting the task',
+  );
+  log.info({ data }, 'publish with TopicArn');
+
+  command = new PublishCommand({
+    Message: 'message to be sent',
+    TopicArn: topicArn + '-unexistent',
+  });
+  try {
+    data = await snsClient.send(command);
+    throw new Error('expected NotFoundException error');
+  } catch (err) {
+    log.info({ data }, 'publish with non existent TopicArn');
+    const statusCode = err && err.$metadata && err.$metadata.httpStatusCode;
+    if (statusCode !== 404) {
+      throw err;
+    }
+  }
+
+  command = new PublishCommand({
+    Message: 'message to be sent',
+    TopicArn: 'arn:aws:sns:us-west-2:111111111111:ignore-name',
+  });
+  try {
+    data = await snsClient.send(command);
+    throw new Error('expected NotFoundException error');
+  } catch (err) {
+    log.info({ data }, 'publish with TopicArn to ignore');
+    const statusCode = err && err.$metadata && err.$metadata.httpStatusCode;
+    if (statusCode !== 404) {
+      throw err;
+    }
+  }
+
+  // TODO: execute a publish with TargetArn in input, combinations of phone and arn?
```

Comment:
Question to the reviewer: should we test if a command combines different properties like `PhoneNumber` and `TopicArn` the span produced has one or the other according to v2 instrumentation?

Reply:
> execute a publish with TargetArn in input

I wonder if that TODO is out of date. There *is* a test above using `TopicArn` isn't there (line 114)?

> should we test if a command combines different properties like `PhoneNumber` and `TopicArn`

Up to you. It is clearly incorrect user behaviour, so it isn't a very high prio to test it. Can you tell from the AWS docs if that is even allowed? If you try it, does the aws-sdk throw an error?

Reply:
> I wonder if that TODO is out of date.

yes it is :) 

> Can you tell from the AWS docs if that is even allowed? If you try it, does the aws-sdk throw an error?

I did not run tests on our AWS env. Will check what happens

Reply:
Did a test by creating a command with `PhoneNumber`, `TopicArn` and `TargetArn` set to a value

```js
command = new PublishCommand({
    Message: 'message to be sent',
    PhoneNumber: '+34555555555',
    TopicArn: 'arn:aws:sns:us-east-2:000000000000:elasticapmtest-topic-3',,
    TargetArn: 'arn:aws:sns:us-east-2:000000000000:elasticapmtest-target-3',
});
data = await snsClient.send(command);
```

And the request build by the SNS client contains the following data
```
req HttpRequest {
  method: 'POST',
  hostname: 'localhost',
  port: 4566,
  query: {},
  headers: {
    'content-type': 'application/x-www-form-urlencoded',
    'content-length': '585',
    host: 'localhost:4566',
    'x-amz-user-agent': 'aws-sdk-js/3.379.1',
    'user-agent': 'aws-sdk-js/3.379.1 ua/2.0 os/darwin#22.6.0 lang/js md/nodejs#18.16.0 api/sns#3.379.1',
    'amz-sdk-invocation-id': '4f1c8035-bc5f-4077-a9f7-faaf558bffc7',
    'amz-sdk-request': 'attempt=1; max=3',
    'x-amz-date': '20230920T075834Z',
    'x-amz-content-sha256': '164ce271603db8027709a51f7f4e2cf48396c3bdded26bbe9b101b418aec6f7b',
    authorization: 'AWS4-HMAC-SHA256 Credential=fake/20230920/us-east-2/sns/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-request;content-length;content-type;host;x-amz-content-sha256;x-amz-date;x-amz-user-agent, Signature=b02123b2aab4181e6e70ae553259ae5e49d64c05a4fb25c682dd82bfb07e3900'
  },
  body: 'TopicArn=arn%3Aaws%3Asns%3Aus-east-2%3A000000000000%3Aelasticapmtest-topic-3&TargetArn=arn%3Aaws%3Asns%3Aus-east-2%3A000000000000%3Aelasticapmtest-target-3&PhoneNumber=%2B34555555555&Message=message%20to%20be%20sent&MessageAttributes.entry.1.Name=traceparent&MessageAttributes.entry.1.Value.DataType=String&MessageAttributes.entry.1.Value.StringValue=00-782f0d159bf47dcf19422b56fd07888e-bdb3d2ba6d01dca1-01&MessageAttributes.entry.2.Name=tracestate&MessageAttributes.entry.2.Value.DataType=String&MessageAttributes.entry.2.Value.StringValue=es%3Ds%3A1&Action=Publish&Version=2010-03-31',
  protocol: 'http:',
  path: '/',
  username: undefined,
  password: undefined,
  fragment: undefined
}
```

Notice all 3 fields are present in the encoded body so I guess the backend is deciding what to do. I can think of:
- Message is sent to only one destination and there is an order between them.
- Message is sent to all destinations.
- Variations (maybe I'm thinking too much)

As for now I haven't found in the docs what exactly will happen. I wonder if the 1st option is correct what should be our span payload 🤔

Reply:
I like the idea of showing all the identifiers that are passed in, even if that means a long name.

Does that cause a problem for `snsShouldIgnoreCommand`? Can we have it filter on each of the three names, if all three are specified in a single PublishCommand.

Reply:
> I like the idea of showing all the identifiers that are passed in, even if that means a long name.

Me too. IMO if one field like PhoneNumber "wins" it can hide a faulty one like a non-existing `TopicArn`

so let's try it and see how it looks :)

>  Can we have it filter on each of the three names, if all three are specified in a single PublishCommand?

Sure! There's a 1st approach in [this commit](https://github.com/elastic/apm-agent-nodejs/pull/3532/commits/934c5de24c3299e28c311048d2fcb7a80822c5f2).

I'll have both changes for tomorrow's meeting